### ISSUE TITLE
Support other cloud providers

### ISSUE THREAD

**Issue by mcrute** [mcrute/alpine-ec2-ami#99]
If this image were to become the official image for upstream Alpine, we'd need to support more than just AWS. After talking to the Alpine developers, they consider these cloud providers to currently be high priority for their images (there may be others later).

- [ ] Azure
- [ ] GCP
- [ ] Oracle

Making these work will require open sourcing and significant expansion of the identity broker (doable with some time) as well as a lot of re-tooling work for the release process. It might make more sense for us to move a lot of that into the identity broker and keeping the Python scripts a build-only concern.

This issue exists more to document my thoughts and conversations with upstream, and to collect public opinions than something that can immediately be worked upon.

**Comment by tomalok** [comment]
Started glancing at this yesterday... Packer's `ebs-surrogate` builder doesn't seem to have a direct equivalent in Azure, GCP, and Oracle OCI.  (Not sure it's worth supporting Oracle Classic...)

AWS and Azure both have `...-chroot` builders, but that requires an already-running instance.

It appears that the default for most cloud builders is to launch some already-in-existence image (or run Packer on that instance itself), run some provisioners on it to make some modifications, and then convert it into an image.  (I know this is true for `oracle-oci`, based on Day Job experience.)

One common thing OCI, Azure, and GCP seem to have in common, is that they can _import_ images from other sources.  There are some with instance migration tools for stealing from other clouds, but most can import various virtual disk formats,  qcow2, KVM, VHD, VMDK, etc. and raw disk images.  AWS can also [import into AMIs](https://www.wavether.com/2016/11/import-qcow2-images-into-aws).

Enter QEMU.  For one, `qemu-img` can [convert between these different formats](https://docs.openstack.org/image-guide/convert-images.html).

How about we...
* use Packer's `qemu` builder to launch base Alpine images
* provision, per each cloud's needs
* convert the images (if necessary) and import them (as private images)
* test (repeat above if further tweaks are needed)
* copy from source to available destinations, and mark them public
* refresh YAML/JSON image data stored in some well-known location

**Comment by mcrute** [comment]
That's a really interesting idea! It would be pretty nice to have a single (or at least limited number) of qcow2 images we can use to power the clouds. I personally have a use for this in my test lab too and have a different process for building those images today that could be united into this.

In my extremely limited research it looks like we could also use the AWS image import feature to [import qcow2 images](https://www.wavether.com/2016/11/import-qcow2-images-into-aws) to AWS as well.

**Comment by mcrute** [comment]
In the longer term I was also hoping to provide qcow2 images for Alpine as well so that people running on-prem infrastructure could make use of them.

**Comment by tsarna** [comment]
Supporting other cloud providers is a great goal, but I would have to see it be a blocker for better integration with official Alpine. Don't let the perfect be the enemy of the good and all that.

I do quite a lot at my Day Job with Azure, and I'm happy to help in any way I can. Although, creating custom images is something I've never done.

**Comment by tomalok** [comment]
@tsarna there are some things in the works, but in general the plan is to support other clouds for Alpine 3.14 (π?), so we have some time to sort out the details...  :)

**Comment by mcrute** [comment]
@tsarna 

> but I would have to see it be a blocker for better integration with official Alpine. Don't let the perfect be the enemy of the good and all that.

Indeed. We're not blocking on adding new providers and are planning to release upstream official 3.13 images. But @tomalok is correct, we'll hopefully be adding more providers **after** the transition to upstream.

**Comment by tomalok** [comment]
I've forked this repo over to https://gitlab.alpinelinux.org/tomalok/alpine-cloud-images where I'll be working on this, and other stuff.

I've done a Q&D PoC that should lead to making profile configuration more flexible and autogenerate the build matrix -- and allow us to include "Clouds" (and other things) as definable dimensions.
